lenskit-core/src/main/java/org/lenskit/rerank/GreedyRerankingItemRecommender.java

```diff
@@ -0,0 +1,91 @@
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.rerank;
+
+import it.unimi.dsi.fastutil.longs.LongSet;
+import org.lenskit.api.ItemRecommender;
+import org.lenskit.api.Result;
+import org.lenskit.api.ResultList;
+import org.lenskit.basic.AbstractItemRecommender;
+import org.lenskit.results.Results;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * A hybrid item recommender that uses a greedy re-ranking strategy to allow re-ranking of items iteratively. This
+ * general algorithm is commonly employed to efficiently optimize for set properties of a recommendation list such as
+ * inter-item diveristy.
+ *
+ * This algorithm takes a baseline ranking algorithm, gets the top-n recommendations and re-ranks them iteratively.
+ * To select each recommended item, first a scoring algorithm is ran based on the currently selected recommendations
+ * and each candidate item. The item with the highest score is then added to the recommended list. This process repeates
+ * until enough items are recommended.
+ *
+ * @author Daniel Kluver
+ */
+public class GreedyRerankingItemRecommender extends AbstractItemRecommender {
+    private static final Logger logger = LoggerFactory.getLogger(GreedyRerankingItemRecommender.class);
+    private final ItemRecommender baseRecommender;
+    private final CandidateItemSelector selector;
+
+    @Inject
+    public GreedyRerankingItemRecommender(ItemRecommender baseRecommender, CandidateItemSelector selector) {
+        this.baseRecommender = baseRecommender;
+        this.selector = selector;
+    }
+
+
+    @Override
+    protected ResultList recommendWithDetails(long user, int n, @Nullable LongSet candidateItems, @Nullable LongSet exclude) {
+        List<Result> candidates = baseRecommender.recommendWithDetails(user, -1, candidateItems, exclude);
+        //modifiable copy
+        candidates = new ArrayList<>(candidates);
+        if (n<0) {
+            n = candidates.size();
+        }
+
+        List<Result> results = new ArrayList<>(n);
+        for (int i = 0; i<n; i++) {
+            Result nextItem = selector.nextItem(user, n, results, candidates);
+            if (nextItem == null) {
+                break;
+            } else {
+                //TODO: REVIEW: there should be an easier way to do this.
```

Comment:
Yes, there should. I was just looking to see if Guava offers any help, but they just have `removeAll` and `removeIf`, no `removeFirst`.

This might be a hideous abstraction break, but what if the strategy returned an index instead of a result? That wouldn't let it also do the rescoring :(, but it would make it easy to go say 'delete item X'.